### PR TITLE
Fix decimal division bug

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

**Comment by arhimondr** [comment]
https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/presto-users/A4EdQ69S2Og/4sPgaQCqBwAJ

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 1]
```diff
@@ -1338,7 +1338,7 @@ else if (u21 >= 0) {
             qhat = u21 / (v1 & LONG_MASK);
         }
         else {
-            qhat = divideUnsignedLong(u21, v1);
+            qhat = divideUnsignedLong(u21, v1 & LONG_MASK);
```
Comment (wenleix) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: is this change required by Hacker's Delight 9-3?

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 2]
```diff
@@ -1371,20 +1371,22 @@ else if (u21 >= 0) {
         return (int) qhat;
     }
 
-    private static long divideUnsignedLong(long dividend, int divisor)
+    /**
+     * Hacker's Delight 9-3
```
Comment (wenleix) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: Any referred implementation  of this algorithm ?

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 1 above)
Comment (arhimondr) [reply presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: The algorithm explained in the Hacker's Delight takes unsigned long as a divisor. For this very case it should always be `int`. So technically simple int can also be accepted. But it makes it a little bit confusing. Thus i decided to copy the algorithm verbatim.

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java (hunk 2 above)
Comment (arhimondr) [reply presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: I just copied it from the book =))

File: presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java [hunk 3]
```diff
@@ -1371,20 +1371,22 @@ else if (u21 >= 0) {
...
+            return Long.compareUnsigned(dividend, divisor) < 0 ? 0L : 1L;
         }
-        long unsignedDivisor = divisor & LONG_MASK;
-        long quotient = (dividend >>> 1) / (unsignedDivisor >>> 1);
-        long remainder = dividend - quotient * unsignedDivisor;
-        while (remainder < 0) {
-            remainder += unsignedDivisor;
-            quotient--;
+
+        if (dividend > 0) {
```
Comment (rschlussel) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: why is there a special case for dividend > 0?  I haven't spent time thinking through how the algorithm works, but from the chapter you referenced it doesn't look required.

Comment (arhimondr) [reply presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java]: This is a shortcircut for the most common case, when both, dividend and divisor are non negative (leftmost bit is 0)
